### PR TITLE
MAINT: app writers return not completed if no ID inferrable

### PR DESCRIPTION
**Pull request by GavinHuttley**
## Summary by Sourcery

Handle cases where no identifier can be inferred in app writers by returning a NotCompleted error, update id_from_source type hints to allow None, and add corresponding tests.

Enhancements:
- Allow id_from_source to return None and early-return a NotCompleted error in writers when no identifier is available
- Update writer constructor signatures to accept Callable returning Optional[str]

Tests:
- Add parameterized tests to ensure writers return NotCompleted with ERROR when no unique ID can be inferred
- Add a test to verify get_unique_id returns None for None input

### CODE REVIEW

File: tests/test_app/test_io.py [hunk 1]
```diff
@@ -588,6 +588,25 @@ def uniqid(source):
...
+    ],
+)
+def test_writer_no_unique_id(tmp_dir, writer, data, attr, dstore):
+    writer = get_app(writer, data_store=dstore(tmp_dir))
+    # no unique id possible, so m will be NotCompleted error
+    value = {} if attr == "info" else None
+    setattr(data, attr, value)
+    m = writer(data)
+    assert isinstance(m, NotCompleted)
+    assert m.type == "ERROR"
```
[the remark this answers is not in the document]
Comment (GavinHuttley) [reply tests/test_app/test_io.py:607]: NotCompleted's do have an inferrable ID
